Transacation and Block tests

JRafaelSoares opened the pull request:
Issue #15

JRafaelSoares committed:
Merge pull request #4 from BlockChain4Students/master

Added signatures

JRafaelSoares committed:
fix: remove broken path from keygen and add init module files

JRafaelSoares committed:
feat: add transaction tests
fix: transaction hash content
fix: file open leaks

JRafaelSoares committed:
fix: String format error in transaction

JRafaelSoares committed:
feat: complete transaction tests

JRafaelSoares committed:
Issue #15
feat: Complete Block constructor tests
fix: Block Hashing process
fix: Transaction test format
fix: Blockchain key generation inside constructor

JRafaelSoares committed:
Issue #15
feat: Complete Block testing
test: Add previous hash testing to Block

RafaelAPB commented on blockchain/blockchain_data_structure.py:
```diff
@@ -1,59 +1,101 @@
+            raise Exception("Transaction must have an associated node id")
+
+        # What type should from and to address be? How do we define public keys?
+        # TODO - Add from and to address type checks
```
create issue?

JRafaelSoares replied:
Issue created

RafaelAPB commented on blockchain/blockchain_data_structure.py:
```diff
@@ -1,59 +1,101 @@
+        # TODO - Add from and to address type checks
+
+        # What type we want amount to be? Probably not int
+        # TODO - Add amount type check
```
create issue?
ammount could be float

JRafaelSoares replied:
Ammount changed to float and check added!

RafaelAPB commented on blockchain/blockchain_data_structure.py:
```diff
@@ -1,59 +1,101 @@
+
+        try:
+            verify_sig(h, self.signature, self.node_id)  # Throws error if signature is invalid
+        except Exception:
```
Return more info

JRafaelSoares replied:
Added a better error information (although we cannot know why it actually fails, either by integrity of wrong ID)

RafaelAPB commented on blockchain/blockchain_data_structure.py:
```diff
@@ -1,59 +1,101 @@
+        except:
+            raise Exception("Invalid Transactions")
+
+        # TODO - Add check for previous hash to be empty only for first block
```
Those todos are excellent catches: perhaps add one issue if you are not planning to address them atm

JRafaelSoares replied:
Issues added!

RafaelAPB reviewed (commented):
Fantastic work!

CatarinaPedreira commented:
LGTM ;)

JRafaelSoares committed:
Issue #15
fix: Transaction amount change to float
fix: Transaction test checks for float amount

JRafaelSoares committed:
Issue #15
feat: Add test to transaction amount

JRafaelSoares committed:
Issue #15
fix: Add more explicit error message
